### ISSUE TITLE
Use dependencies to determine target execution order, even when skipping dependencies

### ISSUE THREAD

**Issue by blairconrad** [adamralph/bullseye#112]
This may be an uncommon use case, but I find myself running a local build, and skipping dependencies (because I don't want to run a `restore` target, or something that calculates the next released version number), but executing targets that depend on others, such as a `build` and a `test` target. In such cases, I must take pains to order the targets properly, or they will be executed in the wrong order. Consider:

```c#
Target("build", DoesNothing);                     
Target("subtest1", DependsOn("build"));           
Target("subtest2", DependsOn("build"));           
Target("test", DependsOn("subtest1", "subtest2"));
Target("default", DependsOn("test"));             
```

When I run with Bullseye 2.0.0-rc.2+build.115:

```
λ  dotnet run -s build test
Bullseye: Starting... (build test) (skip dependencies)
Bullseye/build: Starting...
Bullseye/build: Succeeded. (4.57 ms)
Bullseye/test: Succeeded.
Bullseye: Succeeded. (build test) (skip dependencies) (19.6 ms)
D:\Sandbox\try-bullseye

λ  dotnet run -s test build
Bullseye: Starting... (test build) (skip dependencies)
Bullseye/test: Succeeded.
Bullseye/build: Starting...
Bullseye/build: Succeeded. (3.51 ms)
Bullseye: Succeeded. (test build) (skip dependencies) (15.7 ms)
```

But I would've hoped the order of execution would be the same. Just because I don't want to run (unasked-for) dependencies, I don't want to throw my usual target-ordering rules out the window.

@adamralph, if you've interest in this change (or at least are not opposed), I'd be happy to work on a PR.

**Comment by adamralph** [comment]
@blairconrad I'm not sure what I think about this yet, but I'll write down my thoughts so far.

As is stands, the behaviour is quite straightforward. During normal execution (not using `-s`), the targets will be executed in the order stated, with any dependencies executed first. Of course, that may mean that one of the latter targets is executed first if it appears in the dependency graph of an earlier target, but this is simply the dependency graph doing it's job. When running with `-s`, the very notion of dependencies is effectively removed, and the targets are simply executed in the order stated. If we make this change, then we will be kind "half using" the dependencies, when dependencies are explicitly stated to be ignored. I'm concerned that the behaviour will be more difficult to understand. Also, we'd lose the ability to explicitly state the order of execution, albeit only when using `-s`.

**Comment by adamralph** [comment]
It may be worth checking if there is any precedent to this in another tool.

**Comment by blairconrad** [comment]
@adamralph, another spin on https://github.com/adamralph/bullseye/issues/112#issuecomment-425202172 is that right now, the behaviour is inconsistent because targets execution order is coerced so it aligns with the dependency tree, unless `--skip-dependencies` is specified. With my proposed change, the target execution order is always coerced to align with the dependency tree, which is arguably more straightforward. :wink:

It's true that with my change, we'd lose the ability to intentionally run targets in arbitrary order, specifically in a way that flout the dependency graph. (This is not quite true. We could run the build multiple times, each time specifying `--skip-dependencies` and a reduced set of targets.) Me, I think users would never want to run the targets in such a weird way. (Admittedly my originally proposed use case may fall into that category as well.)

However (and now I'm expanding the scope of this issue), this issue also ties into the `--parallel` flag. If we skip dependencies and have a parallel invocation, then the multiple targets specified on the command line run the risk of being run all at the same time when maybe they shouldn't've been. Safer to factor in the dependency graph and ensure that targets that would normally be ancestors of other targets are run earlier in the chain, rather than concurrently.

**Comment by adamralph** [comment]
Thanks @blairconrad, some good food for thought there. I'll continue to chew on it. 💭

**Comment by adamralph** [comment]
@blairconrad here's another angle: in [your case](https://github.com/FakeItEasy/FakeItEasy/pull/1456#issuecomment-425124446), `versionInfoFile` isn't actually a dependency of `build`. The fact that you usually skip `versionInfoFile` proves that. Only in some cases, do you _choose_ to run `versionInfoFile`, e.g. when preparing a release on the build server. Theoretically you only have to run that target when you are actually preparing a package to release, e.g. when building a tag. In that case, you'd want to ensure that `versionInfoFile` runs _before_ `build`, but it's more accurate to say that it's a "soft dependency". FAKE actually [provides this very feature](http://fake.build/core-targets.html#Soft-dependencies).

That said, one thing I want to try and avoid if possible is the building, and analysis of a dependency graph. In the current implementation, Bullseye doesn't actually know what the dependency graph is ahead of execution. Rather, it loops through the requested targets and "dumbly" runs (or waits for current execution to finish for) each dependency recursively. I guess the soft dependency feature would still require the building of a dependency graph.

So that's my main pushback: the cost of swapping the current simplistic execution model for a richer one which builds and analyses the dependency graph has to be outweighed the benefit provided. Given that we have arguments that both the current behaviour and the proposed behaviour could be regarded as "correct" and intuitive, I'm having a hard time committing to paying the cost of the added complexity.

**Comment by blairconrad** [comment]
Good point about `versionInfoFile` being a soft dependency, @adamralph. I'd not encountered the term before. It's a useful concept. So useful that I really want to create a "support soft dependencies" issue now, since it would be nice to formalize the pattern we'd been ad hocking in our build script.

I understand the desire to not complicate the code. Thanks for considering the feature. Sounds like we're closing this?

**Comment by blairconrad** [comment]
… only I may have a solution that does not completely disrupt everything. PR (for examination/discussion/interest) incoming.

**Comment by adamralph** [comment]
#118 was to the wrong branch.

**Comment by adamralph** [comment]
Closed in #119.

**Comment by adamralph** [comment]
@blairconrad no, it was my fault. You raised #114 to `master`, which was correct. But master has moved on since 2.0.0 RC2, so #118 was my attempt to cherry-pick the commit back `release-2.0` to create 2.0.0 RC3 (following [Release Flow](https://docs.microsoft.com/en-us/azure/devops/learn/devops-at-microsoft/release-flow)).
